src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1090,36 +1252,166 @@ private void CompileSourceToAssembly(string source)
+                    if (diagnisticRecord.IsSuppressed ||
+                        String.Equals(diagnisticRecord.Id, "CS2008", StringComparison.InvariantCulture) ||
+                        String.Equals(diagnisticRecord.Id, "CS1562", StringComparison.InvariantCulture) ||
+                        String.Equals(diagnisticRecord.Id, "BC2008", StringComparison.InvariantCulture))
+                    {
+                        continue;
+                    }
+
+                    IsError |= diagnisticRecord.Severity == DiagnosticSeverity.Error || (diagnisticRecord.IsWarningAsError && diagnisticRecord.Severity == DiagnosticSeverity.Warning);
+
+                    string errorText = BuildErrorMessage(diagnisticRecord);
+
+                    if (diagnisticRecord.Severity == DiagnosticSeverity.Warning)
+                    {
+                        WriteWarning(errorText);
+                    }
+                    else if (diagnisticRecord.Severity == DiagnosticSeverity.Info)
+                    {
+                        WriteInformation(errorText, s_writeInformationTags);
+                    }
+                    else
+                    {
+                        ErrorRecord errorRecord = new ErrorRecord(
+                            new Exception(errorText),
+                            "SOURCE_CODE_ERROR",
+                            ErrorCategory.InvalidData,
+                            diagnisticRecord);
+
+                        WriteError(errorRecord);
+                    }
+                }
+
+                if (IsError)
+                {
+                    ErrorRecord errorRecord = new ErrorRecord(
+                        new InvalidOperationException(AddTypeStrings.CompilerErrors),
+                        "COMPILER_ERRORS",
+                        ErrorCategory.InvalidData,
+                        null);
+                    ThrowTerminatingError(errorRecord);
+                }
             }
         }
 
-        private AddTypeCompilerError[] GetErrors(ImmutableArray<Diagnostic> diagnostics)
+        private string BuildErrorMessage(Diagnostic diagnisticRecord)
         {
-            return diagnostics.Where(d => d.Severity == DiagnosticSeverity.Error || d.Severity == DiagnosticSeverity.Warning)
-                .Select(d => new AddTypeCompilerError
+            var location = diagnisticRecord.Location;
+
+            if (location.SourceTree == null)
+            {
+                // For some error types (linker?) we don't have related source code.
+                return diagnisticRecord.ToString();
+            }
+            else
+            {
+                var text = location.SourceTree.GetText();
+                var textLines = text.Lines;
+
+                var lineSpan = location.GetLineSpan(); // FileLinePositionSpan type.
+                var errorLineNumber = lineSpan.StartLinePosition.Line;
+                var errorPosition = lineSpan.StartLinePosition.Character;
```

Comment:
errorLineNumber and errorPosition should be part of the error message

Reply:
We get this implicitly in next line.

Reply:
Ah in the `diagnisticMessage`, I see. Probably worth to add comments here, for example to the docstring of this method to show generated error text for the giving input.

Reply:
The API haven't docs. The message format can be changed in Roslyn repo. So I believe we should add short custom comment.